src/docfx/publish/PublishUrlMap.cs

```diff
@@ -132,16 +132,34 @@ private PublishUrlMapItem ResolveOutputPathConflicts(IGrouping<string, PublishUr
             }
 
             // redirection file is preferred than source file
-            var redirection = conflicts.FirstOrDefault(x => x.SourcePath.Origin == FileOrigin.Redirection);
-            if (redirection != null)
+            var redirections = conflicts.Where(x => x.SourcePath.Origin == FileOrigin.Redirection).OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer);
+            var nonRedirections = conflicts.Where(x => x.SourcePath.Origin != FileOrigin.Redirection)
+                .OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer)
+                .Select(x => x.SourcePath.Path.Value).ToList();
+            var redirection = redirections.FirstOrDefault();
+            var redirectionCount = redirections?.Count() ?? 0;
+            var nonRedirectionCount = conflicts.Count() - redirectionCount;
+
+            if (redirectionCount == 0)
             {
-                _errors.Add(Errors.Redirection.RedirectedFileNotRemoved(redirection.SourcePath));
-                return redirection;
+                _errors.Add(Errors.UrlPath.OutputPathConflict(conflicts.First().OutputPath, conflicts.Select(x => x.SourcePath)));
+                return conflicts.OrderBy(x => x.SourcePath.Path, PathUtility.PathComparer).Last();
             }
+            else if (redirectionCount == 1)
+            {
+                _errors.Add(Errors.Redirection.RedirectedFileNotRemoved(nonRedirections));
+                return redirection!;
+            }
+            else
+            {
+                if (nonRedirectionCount > 0)
```

Comment:
Can this be `conflicts.Count() > redirectionCount`? We can save a field declaration